catalog-scripts/idp-catalog-wizard.py

```diff
@@ -0,0 +1,174 @@
+import argparse
+import requests
+import os
+# import yaml
+import re
+from requests.auth import HTTPBasicAuth
+import argparse
+from requests.packages.urllib3.util.retry import Retry
+from requests.adapters import HTTPAdapter
+import subprocess
+
+current_directory = os.path.basename(os.getcwd())
+yaml_content_template = """
+apiVersion: backstage.io/v1alpha1
+kind: Component
+metadata:
+  tags:
+    - java
+    - harness
+    - auto-generated
+  name: {repo_name}
+  annotations:
+    backstage.io/source-location: url:{repo_path}
+spec:
+  type: service
+  system: harness
+  lifecycle: experimental
+  owner: Harness_Account_All_Users
+"""
+
+def get_repositories_api(organization, token):
+    url = f"https://api.github.com/orgs/{organization}/repos"
+    headers = {
+        "Accept": "application/vnd.github.v3+json",
+        "Authorization": f"Bearer {token}",
+        "X-GitHub-Api-Version": "2022-11-28"
+    }
+    response = requests.get(url, headers=headers)
+    return response
```

Comment:
This will only give 100 repos. We need to loop over the headers and then get all the pages.

```
< link: <https://api.github.com/organizations/34780278/repos?page=2>; rel="next", <https://api.github.com/organizations/34780278/repos?page=5>; rel="last"
```

Check the response for `curl -v https://api.github.com/orgs/harness/repos`

Reply:
Here's a function that I wrote sometime ago which can take any github query and fetch all the pages and then stitch the results together
```py

import requests



GITHUB_TOKEN = ""

GLOBAL_HEADERS = {
    "Authorization": f"token {GITHUB_TOKEN}"
}

def fetch_all_pages(query, params=None, headers=None):
    """
    If the query returns paginated results,
    this function recursively fetchs all the results, concatenate and return.
    """
    r = requests.get(query, params=params, headers=headers)
    # print(r)
    if not r.ok:
        raise(Exception("Error in fetch_all_pages", "query : ", query, "r.json() ", r.json()))
    link = r.headers.get('link', None)
    if link is None:
        return r.json()

    if 'rel="next"' not in link:
        return r.json()
    else:
        print("Fetched page")
        next_url = None
        for url in link.split(','):
            if 'rel="next"' in url:
                next_url = url.split(';')[0][1:-1]
        next_url = next_url.strip("<")
        next_url = next_url.strip(">")

        return r.json() + fetch_all_pages(next_url, params=params, headers=headers)

# Find all the issues pep8speaks has commented on
search_results_commenter = fetch_all_pages("https://api.github.com/search/issues?q=commenter:pep8speaks", headers=GLOBAL_HEADERS)

```